src/docfx/build/moniker/MonikersProvider.cs

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal class MonikersProvider
+    {
+        private readonly ConcurrentDictionary<string, Lazy<List<string>>> _cache = new ConcurrentDictionary<string, Lazy<List<string>>>();
+        private readonly List<(Func<string, bool> glob, List<string> monikers)> _rules = new List<(Func<string, bool>, List<string>)>();
+
+        public MonikersProvider(Config config, MonikerRangeParser monikerRangeParser)
+        {
+            foreach (var (key, monikerRange) in config.MonikerRange)
+            {
+                _rules.Insert(0, (GlobUtility.CreateGlobMatcher(key), monikerRangeParser.Parse(monikerRange).ToList()));
```

Comment:
>Insert(0 [](start = 23, length = 8)

why do we insert to the head of the list every time?

Reply:
because the we want to match glob from bottom to top, once match, return it.

---
In reply to: [232154067](https://github.com/dotnet/docfx/pull/3596#discussion_r232154067) [](ancestors = 232154067)

Reply:
`Insert` is an O(n) operation. We can `Add` here and match from last to first

---
In reply to: [232155035](https://github.com/dotnet/docfx/pull/3596#discussion_r232155035) [](ancestors = 232155035,232154067)